### PR TITLE
feat: implement `kwok` cloud provider

### PR DESCRIPTION
**Pull request by vadasambar**
Signed-off-by: vadasambar <surajrbanakar@gmail.com>

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR implements [`kwok`](https://kwok.sigs.k8s.io/) provider (`kwok` is a more performant alternative to kubemark; it doesn't require creation of a separate master/cluster for testing) for scale testing CA to identify and fix issues like https://github.com/kubernetes/autoscaler/issues/5769 

More more info, check the [README](https://github.com/vadasambar/autoscaler/blob/kwok-poc/cluster-autoscaler/cloudprovider/kwok/README.md)

#### Which issue(s) this PR fixes:

There is no issue for this PR. I can create one if needed. 

#### Does this PR introduce a user-facing change?

```release-note
Added: new kwok cloud provider (check https://github.com/kubernetes/autoscaler/blob/kwok-poc/cluster-autoscaler/cloudprovider/kwok/README.md) for more info
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by vadasambar** [comment]
Here's what it looks like in action

image


<details>
<summary>
Here's what the new node looks like:
</summary>

```yaml
suraj@suraj:~/sandbox/go-sandbox/node-templates$ k get no ng-2-62588 -oyaml
apiVersion: v1
kind: Node
metadata:
  annotations:
    cluster-autoscaler.kwok.nodegroup/name: ng-2
    kubeadm.alpha.kubernetes.io/cri-socket: unix:///run/containerd/containerd.sock
    kwok.x-k8s.io/node: fake
    node.alpha.kubernetes.io/ttl: "0"
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: "2023-06-07T05:15:19Z"
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/os: linux
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: kind-worker
    kubernetes.io/os: linux
  name: ng-2-62588
  resourceVersion: "111083"
  uid: 7bd7ff54-0e0a-48b3-bccc-9b3e75658850
spec:
  podCIDR: 10.244.1.0/24
  podCIDRs:
  - 10.244.1.0/24
  providerID: kwok:ng-2-62588
status:
  addresses:
  - address: 172.18.0.3
    type: InternalIP
  - address: kind-worker
    type: Hostname
  allocatable:
    cpu: "12"
    ephemeral-storage: 959786032Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 32781516Ki
    pods: "110"
  capacity:
    cpu: "12"
    ephemeral-storage: 959786032Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 32781516Ki
    pods: "110"
  conditions:
  - lastHeartbeatTime: "2023-06-07T05:16:24Z"
    lastTransitionTime: "2023-06-07T05:16:24Z"
    message: kubelet is posting ready status
    reason: KubeletReady
    status: "True"
    type: Ready
  - lastHeartbeatTime: "2023-06-07T05:16:24Z"
    lastTransitionTime: "2023-06-07T05:16:24Z"
    message: kubelet has sufficient memory available
    reason: KubeletHasSufficientMemory
    status: "False"
    type: MemoryPressure
  - lastHeartbeatTime: "2023-06-07T05:16:24Z"
    lastTransitionTime: "2023-06-07T05:16:24Z"
    message: kubelet has no disk pressure
    reason: KubeletHasNoDiskPressure
    status: "False"
    type: DiskPressure
  - lastHeartbeatTime: "2023-06-07T05:16:24Z"
    lastTransitionTime: "2023-06-07T05:16:24Z"
    message: kubelet has sufficient PID available
    reason: KubeletHasSufficientPID
    status: "False"
    type: PIDPressure
  - lastHeartbeatTime: "2023-06-07T05:16:24Z"
    lastTransitionTime: "2023-06-07T05:16:24Z"
    message: RouteController created a route
    reason: RouteCreated
    status: "False"
    type: NetworkUnavailable
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10247
  images:
  - names:
    - registry.k8s.io/etcd:3.5.6-0
    sizeBytes: 102542580
  - names:
    - docker.io/library/import-2023-03-30@sha256:ba097b515c8c40689733c0f19de377e9bf8995964b7d7150c2045f3dfd166657
    - registry.k8s.io/kube-apiserver:v1.26.3
    sizeBytes: 80392681
  - names:
    - docker.io/library/import-2023-03-30@sha256:8dbb345de79d1c44f59a7895da702a5f71997ae72aea056609445c397b0c10dc
    - registry.k8s.io/kube-controller-manager:v1.26.3
    sizeBytes: 68538487
  - names:
    - docker.io/library/import-2023-03-30@sha256:44db4d50a5f9c8efbac0d37ea974d1c0419a5928f90748d3d491a041a00c20b5
    - registry.k8s.io/kube-proxy:v1.26.3
    sizeBytes: 67217404
  - names:
    - docker.io/library/import-2023-03-30@sha256:3dd2337f70af979c7362b5e52bbdfcb3a5fd39c78d94d02145150cd2db86ba39
    - registry.k8s.io/kube-scheduler:v1.26.3
    sizeBytes: 57761399
  - names:
    - docker.io/kindest/kindnetd:v20230330-48f316cd@sha256:c19d6362a6a928139820761475a38c24c0cf84d507b9ddf414a078cf627497af
    - docker.io/kindest/kindnetd@sha256:c19d6362a6a928139820761475a38c24c0cf84d507b9ddf414a078cf627497af
    sizeBytes: 27726335
  - names:
    - docker.io/kindest/local-path-provisioner:v0.0.23-kind.0@sha256:f2d0a02831ff3a03cf51343226670d5060623b43a4cfc4808bd0875b2c4b9501
    sizeBytes: 18664669
  - names:
    - registry.k8s.io/coredns/coredns:v1.9.3
    sizeBytes: 14837849
  - names:
    - docker.io/kindest/local-path-helper:v20230330-48f316cd@sha256:135203f2441f916fb13dad1561d27f60a6f11f50ec288b01a7d2ee9947c36270
    sizeBytes: 3052037
  - names:
    - registry.k8s.io/pause:3.7
    sizeBytes: 311278
  nodeInfo:
    architecture: amd64
    bootID: ff1e4c75-5e08-49d6-b47e-808d9b298720
    containerRuntimeVersion: containerd://1.6.19-46-g941215f49
    kernelVersion: 5.15.0-72-generic
    kubeProxyVersion: v1.26.3
    kubeletVersion: v1.26.3
    machineID: 57c8e586bd8b4b2b80d28d156ce6ef03
    operatingSystem: linux
    osImage: Ubuntu 22.04.2 LTS
    systemUUID: Ubuntu 22.04.2 LTS
  phase: Running

```
</details>

Here's what it looks like in Grafana (this is from a different iteration of the test)
image

**Comment by elmiko** [comment]
i'm very curious about the comparisons between kwok and kubemark.

also, why not create this as a gRPC based provider?

as a gRPC provider there won't be any limitations/restrictions based on adding code to this repository.

**Comment by qianlei90** [comment]
I think this provider makes contributors debug or develop CA locally easier than kubemark, and maybe it's a lightweight choice for CI/CD. This provider can also be a startup or a tutorials for who want contribute to CA but lack of cloud resources.

**Comment by vadasambar** [comment]
> i'm very curious about the comparisons between kwok and kubemark.

Thanks for the comment :D 

I am working on a load test for https://github.com/kubernetes/autoscaler/issues/5769. I will try to post some data on that tomorrow and link here. I am hoping it shows a more concrete difference between the two of them performance wise. I didn't get time to elaborate more on this. I will try to do that in an upcoming comment. Most of it boils down:
1. I wanted to test issues like https://github.com/kubernetes/autoscaler/issues/5769 without having to spin up resources in a public cloud provider. kubemark needed 100m CPU per fake node (which would be 1 core per [upto 10 nodes](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/scalability_tests.md#test-setup)). I wanted a more lightweight solution. 
    > Currently we're running HollowNode with a limit of 0.09 CPU core/pod and 220MB of memory. However, if we also take into account the resources absorbed by default cluster addons and fluentD running on the 'external' cluster, this limit becomes ~0.1 CPU core/pod, thus allowing ~10 HollowNodes to run per core (on an "n1-standard-8" VM node).

    https://github.com/kubernetes/community/blob/master/contributors/devel/sig-scalability/kubemark-guide.md#starting-a-kubemark-cluster
3. Easier testing like @qianlei90 mentioned
4. I found it easier to install and try (feel free to disagree here :)) 
5. This is a little ambitious but I hope `kwok` can be used as a dry-run option for different cloud providers 


> 
> also, why not create this as a gRPC based provider?
> 
> as a gRPC provider there won't be any limitations/restrictions based on adding code to this repository.

I wanted to be able to scale-test issues around different cloud providers i.e., 4. Implementing the `kwok` provider in-tree means we are closer to the actual implementation of our (I think) most-used cloud providers (adding gRPC communication in between would mean an extra delay which is not there in our in-tree cloud providers). I was hoping for something like this:
1. Implement in-tree provider to cover most of the common use-cases
2. Implement `kwok` provider for `clusterapi` so that we can provision `kwok` nodes using `clusterapi` provider
3. Implement gRPC provider if there is user demand

For 2, [Richard from CAPA is already working on it](https://kubernetes.slack.com/archives/C8TSNPY4T/p1685648610609449).

**Comment by vadasambar** [comment]
@elmiko you might be interested in https://github.com/kubernetes/autoscaler/issues/5769#issuecomment-1590541506

You can create close to 400 nodes in well under 100m of CPU as opposed to kubemark where 100m of CPU per node is recommended.

**Comment by qianlei90** [comment]
Hi @vadasambar, the Kwok provider is valuable for local development. Could we move forward with this PR and make progress towards merging it? Your contribution is greatly appreciated. Thank you!

**Comment by vadasambar** [comment]
> Hi @vadasambar, the Kwok provider is valuable for local development. Could we move forward with this PR and make progress towards merging it? Your contribution is greatly appreciated. Thank you!

Hi @qianlei90. Thank you for showing interest. I am working on this PR everyday but I haven't pushed my changes.

File: cluster-autoscaler/cloudprovider/kwok/kwok_provider.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+	if err != nil {
+		klog.Fatalf("failed to get kubeclient config for external cluster: %v", err)
+	}
+
+	kubeClient := kubeclient.NewForConfigOrDie(restConfig)
+
+	stopCh := make(chan struct{})
+	lister := kube_util.NewNodeLister(kubeClient, func(n *apiv1.Node) bool {
+		return n.GetAnnotations()[KwokManagedAnnotation] == "fake"
+	}, stopCh)
```
Comment (qianlei90) [inline cluster-autoscaler/cloudprovider/kwok/kwok_provider.go:174]: These code is out of date. CA can run out of cluster and the source of rest config should be same with kube client used in main.go:

https://github.com/kubernetes/autoscaler/blob/bafcc0dfaf887975cba31a20f81cd47dddb2a06f/cluster-autoscaler/main.go#L444

I opened a new PR to refactor those code expecting to reuse the rest config: https://github.com/kubernetes/autoscaler/pull/6180

File: charts/cluster-autoscaler/templates/clusterrole.yaml [hunk 2]
```diff
@@ -120,6 +122,7 @@ rules:
     verbs:
       - list
       - watch
+      - get
```
Comment (vadasambar) [inline charts/cluster-autoscaler/templates/clusterrole.yaml:125]: To get `kwok-provider-config` and `kwok-provider-templates` ConfigMap

File: charts/cluster-autoscaler/templates/clusterrole.yaml [hunk 3]
```diff
@@ -42,6 +42,8 @@ rules:
     verbs:
     - watch
     - list
+    - create
```
Comment (vadasambar) [inline charts/cluster-autoscaler/templates/clusterrole.yaml:45]: For creating and deleting nodes.

File: charts/cluster-autoscaler/templates/configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+    # kwok:
+    #   release: v0.3.0
+    # # you can also disable installing kwok in CA code (and install your own kwok release)
+    # kwok:
+    #   install: false (true if not specified)
+--- 
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kwok-provider-templates
```
Comment (vadasambar) [inline charts/cluster-autoscaler/templates/configmap.yaml:45]: This is a sample `kwok-provider-templates` which helps users to get started with the kwok provider immediately (`kwok-provider-config` is in tune with this configmap).

File: cluster-autoscaler/cloudprovider/kwok/kwok_provider.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+// GetNodeGpuConfig returns the label, type and resource name for the GPU added to node. If node doesn't have
+// any GPUs, it returns nil.
+func (kwok *KwokCloudProvider) GetNodeGpuConfig(node *apiv1.Node) *cloudprovider.GpuConfig {
+	return gpu.GetNodeGPUFromCloudProvider(kwok, node)
+}
+
+// Refresh is called before every main loop and can be used to dynamically update cloud provider state.
+// In particular the list of node groups returned by NodeGroups can change as a result of CloudProvider.Refresh().
+// TODO(vadasambar): implement this
+func (kwok *KwokCloudProvider) Refresh() error {
```
Comment (vadasambar) [inline cluster-autoscaler/cloudprovider/kwok/kwok_provider.go:127]: This is not needed for the first release but good to have in the future.

**Comment by vadasambar** [comment]
@x13n mentioning you since you said you were interested in the PR. 
The goal is to get the first release of the kwok provider out so that we can get user feedback. Any sort of review/feedback is welcome. I want to address review/feedback comments in a separate PR (I will create separate issue if needed) unless they block the release.

**Comment by vadasambar** [comment]
TODOs after demo in the sig meeting: 
- [ ] squash all the commits into 1
- [ ] do another pass through of the PR. I had to delete the pod stuck in `Pending` state to get into `Running` state even though node was present (seems like an underlying kwok issue?)

File: cluster-autoscaler/cloudprovider/kwok/README.md [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+configmap:
+  name: kwok-provider-templates
+  key: kwok-config # default: config
+```
+
+By default, kwok provider looks for `kwok-provider-config` ConfigMap. If you want to use a different ConfigMap name, set the env variable `KWOK_PROVIDER_CONFIGMAP` (e.g., `KWOK_PROVIDER_CONFIGMAP=kpconfig`). You can set this env variable in the helm chart using `kwokConfigMapName` OR you can set it directly in the cluster-autoscaler Deployment with `kubectl edit deployment ...`.
+
+### FAQ
+#### 1. What is the difference between `kwok` and `kwok` provider?
+`kwok` is an open source project under `sig-kwok`.
```
[the remark this answers is not in the document]
Comment (vadasambar) [reply cluster-autoscaler/cloudprovider/kwok/README.md:192]: I always thought there was a sig-kwok :open_mouth: Thanks for pointing my mistake :pray: I will update the doc.

File: cluster-autoscaler/cloudprovider/builder/builder_all.go [hunk 7]
```diff
@@ -87,7 +90,10 @@ var AvailableCloudProviders = []string{
 // DefaultCloudProvider is GCE.
 const DefaultCloudProvider = cloudprovider.GceProviderName
 
-func buildCloudProvider(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+func buildCloudProvider(opts config.AutoscalingOptions,
+	do cloudprovider.NodeGroupDiscoveryOptions,
+	rl *cloudprovider.ResourceLimiter,
+	allNodesLister listersv1.NodeLister) cloudprovider.CloudProvider {
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/builder/builder_all.go:96]: Why do you need a nodeLister to be passed? Can we pass something more generic instead (client)?

Comment (vadasambar) [reply cluster-autoscaler/cloudprovider/builder/builder_all.go:96]: Thank you for the comment. 

The intention was to pass only what is needed. A caveat however is, cloudprovider implementers would have to change this if they need more than just node lister and possibly touch other cloudprovider's code to make it compatible as well. 

My initial implementation was based on informer (which is more generic). We can pass it here. I can make that change. Let me know if you have any differing thoughts.

Comment (vadasambar) [reply cluster-autoscaler/cloudprovider/builder/builder_all.go]: Passing in informers now.

**Pull request merged**